index.bs

```diff
@@ -8041,63 +8043,367 @@ ECMAScript Object values.
+
+    Note: If the provided steps return a value of type |T|, this algorithm returns a
+    <code>{{Promise}}&lt;<var ignore>T</var>&gt;</code>.
 </div>
 
-<p class="issue">
-    Include an example of how to write spec text using this term.
-</p>
+<div algorithm>
+    To perform some steps <dfn export for=Promise>upon fulfillment</dfn> of a
+    <code>{{Promise}}&lt;|T|&gt;</code> |promise| given some steps |steps| taking a value of type
+    |T|, perform the following steps:
+
+    1.  [=Upon settling=] of |promise|:
+        *   If |promise| was fulfilled with value |v|, then:
+            1.  Perform |steps| with |v|.
+</div>
+
+<div algorithm>
+    To perform some steps <dfn export for=Promise>upon rejection</dfn> of a
+    <code>{{Promise}}&lt;<var ignore>T</var>&gt;</code> |promise| given some steps |steps| taking
+    an ECMAScript value, perform the following steps:
+
+    1.  [=Upon settling=] of |promise|:
+        *   If |promise| was rejected with reason |r|, then:
+            1.  Perform |steps| with |r|.
+</div>
+
+<div algorithm>
+    To <dfn export for=Promise lt="wait for all|waiting for all">wait for all</dfn> with a [=list=]
+    of <code>{{Promise}}&lt;|T|&gt;</code> values |promises|, with success steps |successSteps|
+    that take a [=list=] of |T| values and failure steps |failureSteps| that take a rejection
+    reason {{any}} value, perform the following steps:
+
+    1.  Let |fullfilledCount| be 0.
+    1.  Let |rejected| be false.
+    1.  Let |rejectionHandlerSteps| be the following steps given |arg|:
+        1.  If |rejected| is true, abort these steps.
+        1.  Set |rejected| to true.
+        1.  Perform |failureSteps| given |arg|.
+    1.  Let |rejectionHandler| be [=!=] [$CreateBuiltinFunction$](|rejectionHandlerSteps|, « »):
+    1.  Let |total| be |promises|'s [=list/size=].
+    1.  If |total| is 0, then:
+        1.  [=Queue a microtask=] to perform |successSteps| given « ».
+        1.  Return.
+    1.  Let |index| be 0.
+    1.  Let |result| be a [=list=] containing |total| null values.
+    1.  [=list/For each=] |promise| of |promises|:
+        1.  Let |promiseIndex| be |index|.
+        1.  Let |fulfillmentHandler| be the following steps given |arg|:
+            1.  Set |result|[|promiseIndex|] to |arg|.
+            1.  Set |fullfilledCount| to |fullfilledCount| + 1.
+            1.  If |fullfilledCount| equals |total|, then perform |successSteps| given |result|.
+        1.  Let |fulfillmentHandler| be [=!=] [$CreateBuiltinFunction$](|fulfillmentHandler|, « »):
+        1.  Perform [$PerformPromiseThen$](|promise|, |fulfillmentHandler|, |rejectionHandler|).
+        1.  Set |index| to |index| + 1.
+</div>
+
+This phrase is useful when you wish to aggregate the result of multiple promises, and react to them
+all together, in the same way that {{Promise/all()|Promise.all()}} functions for JavaScript code.
```

Comment:
Probably the clause "in the same way..." should move to the "get a promise for waiting for all". I think this was just missed when I separated the two operations in the promises guide.